docs/development/CUSTOM_RULES.md

```diff
@@ -0,0 +1,206 @@
+---
+id: custom-rules
+sidebar_label: Custom Rules
+title: Custom Rules
+---
+
+:::important
+You should be familiar with [ESLint's developer guide](https://eslint.org/docs/developer-guide) and [Development > Architecture](./architecture/asts) before writing custom rules.
+:::
+
+As long as you are using `@typescript-eslint/parser` as the `parser` in your ESLint configuration, custom ESLint rules generally work the same way for JavaScript and TypeScript code.
+The main two changes to custom rules writing are:
+
+- [AST Extensions](#ast-extensions): targeting TypeScript-specific syntax in your rule selectors
+- [Typed Rules](#typed-rules): using the TypeScript type checker to inform rule logic
+
+## AST Extensions
+
+`@typescript-eslint/estree` creates AST nodes for TypeScript syntax with names that begin with `TS`, such as `TSInterfaceDeclaration` and `TSTypeAnnotation`.
+These nodes are treated just like any other AST node.
+You can query for them in your rule selectors.
+
+This rule written in JavaScript bans interfaces that start with a lower-case letter:
+
+```js
+export const rule = {
+  create(context) {
+    return {
+      TSInterfaceDeclaration(node) {
+        if (/\p{Lu}/.test(node.id.name[0])) {
+          context.report({
+            messageId: 'uppercase',
+            node: node.id,
+          });
+        }
+      },
+    };
+  },
+  meta: {
+    docs: {
+      category: 'Best Practices',
+      description: 'Interface names should start with an upper-case letter.',
+    },
+    messages: {
+      uppercase: 'Start this name with an upper-case letter.',
+    },
+    type: 'suggestion',
+    schema: [],
+  },
+};
+```
+
+### Node Types
+
+TypeScript types for nodes exist in a `TSESTree` namespace exported by `@typescript-eslint/experimental-utils`.
+The above rule body could be better written in TypeScript with a type annotation on the `node`:
+
+```ts
+import { TSESTree } from '@typescript-eslint/experimental-utils';
+import * as eslint from 'eslint';
+
+export const rule: eslint.Rule.RuleModule = {
```

Comment:
![](https://c.tenor.com/fBvQV_5Lp6UAAAAC/we-dont-do-that-here-black-panther.gif)

The ESLint types are all SUPER wrong for TS development because `@types/eslint` is based on `@types/estree`. `@types/estree` is for the base ESTree spec, so it is missing all the TS node types and everything else.

In fact - my recommendation is for people to never actually import from `'eslint'` directly because of this! Instead we have all the types someone needs exported from `/experimental-utils` which are the same as `@types/eslint`, but setup to align with our infra.

```suggestion
import {TSESLint} from '@typescript-eslint/experimental-utils';

export const rule: TSESLint.RuleModule = {
```

But - even better is if you leverage the `RuleCreator`, as it'll do a bang up job of inferring types of `messageId`s automatically.

```suggestion
import {ESLintUtils} from '@typescript-eslint/experimental-utils';

const createRule = ESLintUtils.RuleCreator(name => `https://example.com/rule/${name}`);

export const rule = createRule({
```